### ISSUE TITLE
High CPU usage

### ISSUE THREAD

**Issue by jpop32** [ava-labs/avalanchego#258]
**Describe the bug**
CPU pegged on two cores ('top' showing 100-200% CPU usage for ava process). 

**To Reproduce**
Upgraded from 0.5.5 to 0.5.7 and left it running for some 24hrs.

**Expected behavior**
CPU is usually at 1-5% usage. After restarting the node, it went back to normal.

**Operating System**
Ubuntu Server 20.04, 64bit, on Raspberry Pi4b.

**Additional context**
Node built from [0.5.7 tag](https://github.com/ava-labs/gecko/commit/e30f00c0f2254852ee2a68c49e07d8a99fbd6b95)
Node was staking at the time. NodeID=6gRx5vFhuMTDtaRYP82ZSGuBq7eR5s3jw
Ran CPU profile on the node for a couple of seconds before restarting:
[cpu.zip](https://github.com/ava-labs/gecko/files/4833623/cpu.zip)

By submitting this issue I agree to the [Terms and Conditions](https://files.avalabs.org/legal/bug_hunt_q2_2020_terms.pdf) of the [Developer Accelerator Program](https://avalabs.org/dap).

**Comment by jpop32** [comment]
And again. Went away for the weekend, when I came back, again high CPU load on the ava process.

Here's the CPU profile. 
[cpu.zip](https://github.com/ava-labs/gecko/files/4842566/cpu.zip)

**Comment by jpop32** [comment]
Looks like it happens regularly after some 24 hours of running.

Got all the profiles this time. 
[profile.zip](https://github.com/ava-labs/gecko/files/4847568/profile.zip)

**Comment by danlaine** [comment]
Hey @jpop32 thanks for reporting this and for the profiles. Much appreciated!

You mentioned on Discord that your ISP gives you a new public IP every day, right?
Have you noticed a connection between getting a new IP and the CPU usage increasing?

**Comment by jpop32** [comment]
That indeed might be it. I'll power cycle my router (after everyone has gone to sleep, of course), to trigger IP change to see if that triggers it.

**Comment by jpop32** [comment]
Looks like that's not it. I forced connection IP change twice, and CPU usage remained ok, low single digits. Going to leave it running, so we'll see tomorrow.

I'm going to install Prometheus+Grafana, to better monitor node health, that might be helpful.

**Comment by el33th4x0r** [comment]
Hi Jovica,

Could you please check your IP address (https://www.whatsmyip.org/) to see
if it changes? Sometimes a reboot won't trigger a change.

Thank you once again, very much appreciate your help in debugging this
issue!
- egs

**Comment by jpop32** [comment]
I checked on the router each time, and wrote the IP down before rebooting it, it has changed. It's been 8 hours now, and it's still fine. Yesterday I configured port forwarding for the staking port (9651), that might have had an effect. I'll investigate further this afternoon.

**Comment by jpop32** [comment]
Played with it some more. Doesn't look like IP changes are what's causing it. Forced IP change two more times, second time with port forwarding disabled. CPU does increase while re-syncing and acquiring peers, but then it drops back to usual.

Since the node is on wifi, restarting the router also restarts the net connection on the machine. When the ISP drops the link and changes the IP that doesn't happen, could that play a role? Gonna leave it running to see what happens when my ISP drops the link tomorrow.

BTW, I've set up Prometheus and Grafana, so if it happens again, I can pull out anything that might be useful: gecko, go, net, node or process metrics. LMK.

**Comment by danlaine** [comment]
@jpop32 Is your computer goes to sleep when you leave it for a long time, and the ava process stops executing?

**Comment by jpop32** [comment]
No, I don't think so. It's running Ubuntu Server 20.04, so it's built to run 24/7. My node has been running on it every since Denali challenge ended, and it worked fine up until the 0.5.7 upgrade.

**Comment by jpop32** [comment]
Update: It's been more than 24 hours since last high-CPU usage incident. But, coincidentally, last night my staking time expired, I'm no longer a validator. So there's reason to believe it only happens while you're a validator.

Gonna re-stake now, to confirm.

**Comment by jpop32** [comment]
An update. Things have been going smooth for the last week or so. During that time I've monitored my IP, and it hasn't changed (my ISP moves in mysterious ways). Went away for the weekend, and came back to having high CPU usage. In the meantime I've set up monitoring (by logging gecko metrics API), and this was the situation:

<img width="428" alt="Annotation 2020-07-12 195743" src="https://user-images.githubusercontent.com/6398640/87253839-943a6600-c47e-11ea-88cc-5eef0419f282.png">

Since I don't believe (and explorer supports the conclusion) denali has been processing 2500 vertices every 15 seconds, this is a manifestation of the bug. IP hasn't changed before or after high usage started. But my staking period has ended on the 9th. so this happened while running as a regular node, not a validator.

Profiles: [profile.zip](https://github.com/ava-labs/gecko/files/4909202/profile.zip)
